### PR TITLE
[None][feat] Add test for speculative rejection sampler (2-model)

### PR DESCRIPTION
**Pull request by IzzyPutterman**
## Summary by CodeRabbit

* **New Features**
  * Added support for deterministic sampling using seeded random generators, ensuring consistent results across multi-GPU setups.
  * Introduced temperature control for top-p sampling, allowing more flexible sampling strategies.
  * Implemented probabilistic acceptance of draft tokens based on model confidence, improving speculative decoding.

* **Improvements**
  * Enhanced handling of draft logits and probabilities for more accurate and controlled token generation.

* **Tests**
  * Added unit tests validating the rejection sampling mechanism to ensure accurate probabilistic token acceptance.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by mikeiovine — COMMENTED** [review]
Can we ban using ngram when sampling is enabled? We'll need to do more work to enable it

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 1]
```diff
@@ -179,12 +215,14 @@ def sampling_strategies(requests: Iterable[LlmRequest]) -> list[Strategy]:
     return [request_strategy(req) for req in requests]
 
 
-def sample(strategy: Strategy, logits: torch.Tensor):
+def sample(strategy: Strategy,
+           logits: torch.Tensor,
+           generator: torch.Generator = None):
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:220]: Type should be `Optional[torch.Generator]`

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 2]
```diff
@@ -239,6 +277,24 @@ def __init__(self, args: Args):
         # So, we temporarily exit inference mode.
         with torch.inference_mode(False):
             self.store = self.create_store()
+            # Initialize seed for multi-GPU consistency
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:280]: Comment has wrong indentation

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 3]
```diff
@@ -302,20 +358,70 @@ def handle_logits(self, request: LlmRequest, state: SampleState, *,
...
+                    break
+
+                num_accepted += 1
+                new_token = add_token(request,
+                                      new_tokens,
+                                      beam=self.BEAM,
+                                      step=num_accepted)
+                if self._handle_stop_criteria(request, new_token):
+                    break
+            return num_accepted
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:382]: Can we separate the two branches into functions: `_process_draft_tokens_greedy` and `_process_draft_tokens_rejection_sampling`?

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 4]
```diff
@@ -157,8 +164,36 @@ def greedy_search_sampling_batch(logits):
     return next_tokens, softmax
 
 
+def get_rejected_indices(draft_probs: torch.Tensor, target_probs: torch.Tensor,
+                         generator: torch.Generator, draft_tokens: list[int]):
+
+    p = draft_probs[torch.arange(len(draft_tokens)), draft_tokens]
+    q = target_probs[:-1]
+    q = q[torch.arange(len(draft_tokens)), draft_tokens]
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:172]: `q.shape[0] == len(draft_tokens)` at this point right? It can be simplified to just this

```python
q = q[:, draft_tokens]
```

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 5]
```diff
@@ -302,20 +358,70 @@ def handle_logits(self, request: LlmRequest, state: SampleState, *,
...
+            stop = False
+            if rejected_indices.numel() == 0:
+                num_initially_accepted = len(request.py_draft_tokens)
+                sample_last = False
+            else:
+                num_initially_accepted = rejected_indices[0].item()
+            num_accepted = num_initially_accepted
+            for i in range(num_accepted):
+                new_token = request.py_draft_tokens[i]
+                new_tokens[i, request.seq_slot, self.BEAM] = new_token
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:404]: No need to mutate `new_tokens` here

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 6]
```diff
@@ -302,20 +358,70 @@ def handle_logits(self, request: LlmRequest, state: SampleState, *,
...
+                new_tokens[num_accepted, request.seq_slot,
+                           self.BEAM] = new_token
+                request.add_new_token(new_token, self.BEAM)
+                stop = self._handle_stop_criteria(request, new_token)
+            elif not stop and not sample_last:
+                new_token = add_token(request,
+                                      new_tokens,
+                                      beam=self.BEAM,
+                                      step=num_accepted)
+                stop = self._handle_stop_criteria(request, new_token)
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:422]: Might be clearer to just have this above this block of code:
```python
if stop:
    return num_accepted
```

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 7]
```diff
@@ -302,20 +358,70 @@ def handle_logits(self, request: LlmRequest, state: SampleState, *,
...
+                if self._handle_stop_criteria(request, new_token):
+                    break
+            return num_accepted
+
+        else:
+            sampling_strategy = request_strategy(request)
+            generator = self.get_generator(request.py_draft_logits.device)
+            _, draft_probs = sample(sampling_strategy,
+                                    request.py_draft_logits[0],
+                                    generator=generator)
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/pyexecutor/sampler.py:389]: I'm totally in favor just getting this working first and then optimizing it later, but we should be able to cache the draft probs in the drafter. Not sure if it makes a difference. We can prioritize having simpler code for now

File: tests/unittest/_torch/speculative/test_torch_rejection_sampling.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    expected_prob = target_probs[0].squeeze().numpy()
+
+    # KL Divergence check
+    kl_divergence = entropy(expected_prob, sampled_tokens)
+    kl_divergence_regular = entropy(expected_prob, sampled_regular)
+    assert abs(kl_divergence - kl_divergence_regular) < 0.01
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (mikeiovine) [inline tests/unittest/_torch/speculative/test_torch_rejection_sampling.py:51]: Needs `import unittest`

File: tests/unittest/_torch/speculative/test_torch_rejection_sampling.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
+import numpy as np
+import torch
+from scipy.stats import entropy
+
+from tensorrt_llm._torch.pyexecutor.sampler import (get_rejected_indices,
+                                                    sample_rejected)
+
+
+def test_get_rejected_indices():
```
Comment (mikeiovine) [inline tests/unittest/_torch/speculative/test_torch_rejection_sampling.py:9]: It would be great if we had a way to test the sampler directly instead of testing the algorithm through the helper functions. Unfortunately working with the `LlmRequest` objects outside of `PyExecutor` is a huge pain. This should be good enough for now.

**Review by mikeiovine — COMMENTED** [review]
Thank you @IzzyPutterman! The comments are mostly nitpicks.

**Comment by IzzyPutterman** [comment]
Will wait for GPT OSS PR to land, then update with these changes.

File: tensorrt_llm/_torch/pyexecutor/sampler.py (hunk 4 above)
Comment (IzzyPutterman) [reply tensorrt_llm/_torch/pyexecutor/sampler.py:172]: we want the final shape to be 1 x draft _tokens, wouldnt that return draft_tokens x draft_tokens?

Comment (mikeiovine) [reply tensorrt_llm/_torch/pyexecutor/sampler.py:219]: A bit confused. Isn't the first dimension `len(draft_tokens)` in the current version?

**Review by mikeiovine — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/sampler.py (hunk 5 above)
Comment (mikeiovine) [reply tensorrt_llm/_torch/pyexecutor/sampler.py]: Yep, my bad

File: tensorrt_llm/_torch/pyexecutor/sampler.py (hunk 4 above)
Comment (IzzyPutterman) [reply tensorrt_llm/_torch/pyexecutor/sampler.py:219]:
```
>>> import torch
>>> a = torch.randn(3, 100)
>>> a[:, [0, 1, 2]]
tensor([[ 0.0327,  0.4760,  0.1679],
        [ 0.4568, -0.2794,  0.4780],
        [-1.9470, -0.4793,  1.7912]])
>>> a[[0,1,2], [0,1,2]]
tensor([ 0.0327, -0.2794,  1.7912])
```
